### PR TITLE
Import coilset details from eqdsk if available

### LINKED ISSUE #645: Using CoilSet.from_eqdsk mis-classifies coils
## Describe the bug

The [CoilSet.from_group_vecs](https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/develop/bluemira/equilibria/coils.py#L1770) method (called by [CoilSet.from_eqdsk](https://github.com/Fusion-Power-Plant-Framework/bluemira/blob/develop/bluemira/equilibria/coils.py#L1746)) uses a simple heuristic to distinguish between PF coils and CS coils: if `dx` and `dz` are equal then they are PF coils, otherwise they are CS. Presumably this has been sufficient when working with conventional Tokamak designs but it is not reliable for Spherical Tokamaks. 

## Steps to reproduce

I encountered this when trying to design some new behaviour for the STEP code (making it possible to read in old runs). This code is not public. A JSON serialisation of the original coils is given below:
```json
{
    "PF_1.1": {
        "x": 1.5,
        "z": 8.787573575105338,
        "dx": 0.175,
        "dz": 0.5,
        "current": 5894660.784269055
    },
    "PF_1.2": {
        "x": 1.5,
        "z": -8.787573575105338,
        "dx": 0.175,
        "dz": 0.5,
        "current": 5894660.784269055
    },
    "PF_2.1": {
        "x": 1.5,
        "z": 11.307573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 6791296.337239938
    },
    "PF_2.2": {
        "x": 1.5,
        "z": -11.307573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 6791296.337239938
    },
    "PF_3.1": {
        "x": 12.802467952311037,
        "z": 11.807573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 32920322.657766376
    },
    "PF_3.2": {
        "x": 12.802467952311037,
        "z": -11.807573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 32920322.657766376
    },
    "PF_4.1": {
        "x": 12.802467952311037,
        "z": 6.8,
        "dx": 0.25,
        "dz": 0.4,
        "current": -41363029.116069205
    },
    "PF_4.2": {
        "x": 12.802467952311037,
        "z": -6.8,
        "dx": 0.25,
        "dz": 0.4,
        "current": -41363029.116069205
    },
    "PF_5.1": {
        "x": 15.120639645587653,
        "z": 1.7,
        "dx": 0.35,
        "dz": 0.5,
        "current": 8144992.116116044
    },
    "PF_5.2": {
        "x": 15.120639645587653,
        "z": -1.7,
        "dx": 0.35,
        "dz": 0.5,
        "current": 8144992.116116044
    },
    "CS_1": {
        "x": 0.1225,
        "z": 4.205,
        "dx": 0.022,
        "dz": 2.0525,
        "current": 0
    },
    "CS_2": {
        "x": 0.1225,
        "z": 0.0,
        "dx": 0.022,
        "dz": 2.0525,
        "current": 0
    },
    "CS_3": {
        "x": 0.1225,
        "z": -4.205000000000001,
        "dx": 0.022,
        "dz": 2.0525,
        "current": 0
    }
}
```
After read-in, they were serialised as follows:
```json
{
    "CS_17": {
        "x": 1.5,
        "z": 8.787573575105338,
        "dx": 0.175,
        "dz": 0.5,
        "current": 5894660.784269055
    },
    "CS_18": {
        "x": 1.5,
        "z": -8.787573575105338,
        "dx": 0.175,
        "dz": 0.5,
        "current": 5894660.784269055
    },
    "CS_19": {
        "x": 1.5,
        "z": 11.307573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 6791296.337239938
    },
    "CS_20": {
        "x": 1.5,
        "z": -11.307573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 6791296.337239938
    },
    "CS_21": {
        "x": 12.802467952311037,
        "z": 11.807573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 32920322.657766376
    },
    "CS_22": {
        "x": 12.802467952311037,
        "z": -11.807573575105339,
        "dx": 0.25,
        "dz": 0.4,
        "current": 32920322.657766376
    },
    "CS_23": {
        "x": 12.802467952311037,
        "z": 6.8,
        "dx": 0.25,
        "dz": 0.4,
        "current": -41363029.116069205
    },
    "CS_24": {
        "x": 12.802467952311037,
        "z": -6.8,
        "dx": 0.25,
        "dz": 0.4,
        "current": -41363029.116069205
    },
    "CS_25": {
        "x": 15.120639645587653,
        "z": 1.7,
        "dx": 0.35,
        "dz": 0.5,
        "current": 8144992.116116044
    },
    "CS_26": {
        "x": 15.120639645587653,
        "z": -1.7,
        "dx": 0.35,
        "dz": 0.5,
        "current": 8144992.116116044
    }
}
```
(There are also three coils missing but this appears to be due to a problem with the STEP reactor class itself creating coils separately from the EQDSK.)

## Expected behaviour

The names of the coils in JSON above should be the same.

## Environment (please complete the following information)

- OS: Ubuntu 20.0
- Branch: cmacmackin/build-refactor-demonstration (also applies on develop)

## Additional context

This is not an urgent bug for me as I can reconstruct the coils directly from another serialisation produced by the STEP reactor code.

### LINKED ISSUE #3294: Integrate coil names and types from eqdsk into CoilSet
## Description of issue / requirement to address

#3242 introduced saving and loading coil names and types to/from an eqdsk.json.

These changes must get integrated into the CoilSet created when loading an eqdsk using the `Equilibrium` class.

### PR DESCRIPTION
**Pull request by je-cook**
## Linked Issues

Closes #3294 Closes #645

#645 is probably an unfixable problem for raw eqdsk but this sidesteps that problem for json

## Description

Imports coil names into equilibrium objects

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `pre-commit run --from-ref develop --to-ref HEAD`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/equilibria/coils/_grouping.py [hunk 1]
```diff
@@ -308,10 +309,21 @@ def from_group_vecs(cls, eqdsk: EQDSKInterface) -> CoilGroup:
...
+                return lst[idx]
+            except IndexError:
+                return default
+
         for i in range(eqdsk.ncoil):
             dx = eqdsk.dxc[i]
             dz = eqdsk.dzc[i]
-            if abs(eqdsk.Ic[i]) < I_MIN:
+            cn = _get_val(eqdsk.coil_names, i)
+            ct = v if (v := _get_val(eqdsk.coil_types, i)) is None else CoilType(v)
```
Comment (oliverfunk) [inline bluemira/equilibria/coils/_grouping.py:325]: Just to clarify here, if `_get_val(eqdsk.coil_types, i)` results in None, ct will = None right?

File: bluemira/equilibria/coils/_grouping.py [hunk 2]
```diff
@@ -322,18 +334,20 @@ def from_group_vecs(cls, eqdsk: EQDSKInterface) -> CoilGroup:
...
+            elif ct is CoilType.CS or (dx != dz and ct is None):  # Rough and ready
                 cscoils.append(
                     Coil(
                         eqdsk.xc[i],
                         eqdsk.zc[i],
                         current=eqdsk.Ic[i],
                         dx=dx,
                         dz=dz,
-                        ctype="CS",
+                        ctype=ct or CoilType.CS,
```
Comment (oliverfunk) [inline bluemira/equilibria/coils/_grouping.py:349]:
```suggestion
                        ctype=CoilType.CS,
```

Can only be `CS` here right?

Comment (oliverfunk) [reply bluemira/equilibria/coils/_grouping.py:349]: Or `None` but I think we want the ctype to be `CS` here

File: bluemira/equilibria/coils/_grouping.py (hunk 1 above)
Comment (je-cook) [reply bluemira/equilibria/coils/_grouping.py:325]: yep, avoiding passing None into coiltype

File: bluemira/equilibria/coils/_grouping.py [hunk 3]
```diff
@@ -343,7 +357,8 @@ def from_group_vecs(cls, eqdsk: EQDSKInterface) -> CoilGroup:
                     current=eqdsk.Ic[i],
                     dx=dx,
                     dz=dz,
-                    ctype="PF",
+                    ctype=ct or CoilType.PF,
```
Comment (oliverfunk) [inline bluemira/equilibria/coils/_grouping.py:360]: Do we want to allow other ctype's here other than PF?

Comment (je-cook) [reply bluemira/equilibria/coils/_grouping.py:360]: yes dummy coils now or any future coil type could be named here.

**Review by oliverfunk — APPROVED** [review]
Super great!

**Pull request merged**